airflow/datasets/__init__.py

```diff
@@ -271,6 +306,20 @@ def iter_datasets(self) -> Iterator[tuple[str, Dataset]]:
                 yield k, v
                 seen.add(k)
 
+    def iter_dag_deps(self, *, source: str, target: str) -> Iterator[DagDependency]:
+        """
+        Iterate dataset, dataset aliases and their resolved datasets  as dag dependency.
+
+        :meta private:
+        """
+        dag_deps: set[DagDependency] = set()
```

Comment:
Is the order of DAG dependencies significant? If not, I think we can deduplicate at where this function is called instead. If we have aliases nested in conditions (or just nested conditions), each nesting would have a set doing deduplication, which isn’t efficient IMO.

Or, do we need to deduplicate dependencies at all…?

Reply:
> Is the order of DAG dependencies significant? If not, I think we can deduplicate at where this function is called instead. 

I think not. but I'll test out.

> Or, do we need to deduplicate dependencies at all…?

I would probably a +0? e.g., `scheudle=[Datset("test-1"), Dataset("test-1")]` should we generate 2 `Dataset("test-1")`? On the other hand, it's also what users want 🤔  I feel it's more like a UI thing and would like to know how @bbovenzi  think

Reply:
oh, in the current dateset_dependecnies endpoint, I think it'll be filtered out anyway. so we can also defer the deduplication later as well

Reply:
I feel that not deduplicating might make it even more complicated.

Reply:
Can you elaborate?

Reply:
Sure.

currently, a DAG with `schedule=[Dataset("1"), Dataset("1")]` will generate the following graph.


```mermaid
flowchart LR
    dataset1 --> DAG1
```

By not deduplicate, my understanding is we want to have something like this. Then we'll have to mark that the frist `Dataset("1")` and the second `Dataset("1")` are different. It's not undoable. Just not sure whether it makes sense.

```mermaid
flowchart LR
    dataset1 --> DAG1
    dataset1 --> DAG1
```

Reply:
AFAIK, we do not have `DatasetAll` and `DatasetAny` in the graph. so `schedule=Dataset("1") & Dataset("1") | Dataset("1")` would be something similar to the above comment.

Reply:
In this particular situation, I wonder if we should just deduplicate in DatasetAll and DatasetAny. I anticipate there might be other places we should dedup, but at least for those there’s a simple math solution.

Reply:
What I'm thinking now is whether there's a situation we really don't want deduplication. Or is this more like a performance issue?

Reply:
For DatasetAll and DatasetAny specifically, I don’t think there’s a situation we need to keep the original user input. I’d even argue it’s an oversight I did not implement dedup in `__and__` and `__or__` in the initial release.

There are cases where we can’t dedup in all/any but need further cleanup in the UI though, say `("a" | "b") & ("a" | "c")`. This cannot be simplified, but we only want one line to `a` instead of two. But that’s a UI consideration that should only be done either in the API layer or frontend (React).

Reply:
> say ("a" | "b") & ("a" | "c"). This cannot be simplified, but we only want one line to a instead of two. 

I guess I'm a bit confused. What do you expect the UI to look like in this situation? It would look something like that.

```mermaid
flowchart LR
    a --> DAG
    b --> DAG
    c --> DAG
```

If we don't duplicate (for cases that can't actually be dedup), I would probably want something like this. 

```mermaid
flowchart LR
    a --> Or1
    b --> Or1
    a --> Or2
    c --> Or2

    Or1 --> And
    Or2 --> And

    And --> DAG
```

Reply:
By “now” above you mean before this PR right? That’s what I would expect the graph to look like. So we still need some dedup logic when rendering the UI, regardless what we do in all/any. But I want those to be strictly in the UI layer (i.e. somewhere in API or frontend), not `iter_dag_dependencies`.

Reply:
Got your point! Yep, it's no longer in `iter_dag_dependencies`. But it still happens in `airflow/serialization/serialized_objects.py`. So I think it's still not ` strictly in the UI layer (i.e. somewhere in API or frontend),` But in the "current" (before this PR) version, we're using `iter_datasets` to [detect_dag_dependencies](https://github.com/apache/airflow/blob/9ec9eb79a0cc845d86e7380c73269d2ee1d3c210/airflow/serialization/serialized_objects.py#L963). On the other hand, the [outlet part](https://github.com/apache/airflow/blob/9ec9eb79a0cc845d86e7380c73269d2ee1d3c210/airflow/serialization/serialized_objects.py#L946-L955) did not deduplicate.

I think what you suggest is better. (let the API layer do the deduplication job, I'll probably do it [here](https://github.com/apache/airflow/blob/9ec9eb79a0cc845d86e7380c73269d2ee1d3c210/airflow/www/views.py#L3498)), but would like to check whether changing this behavior (which affect serialized DAG) is actually ok. Thanks!

Reply:
The above is fine for now. But yes, I think we should show the conditions for triggering a DAG in the graph. That is separate from showing dataset aliases at all.

Reply:
@uranusjr I just created https://github.com/apache/airflow/pull/40984 to verify that deduplication is happening in the main branch and would like to confirm it's ok to change that behavior in this branch.

Reply:
@uranusjr I've changed the serialization behavior and moved the deduplication to the API layer.